core-concepts/modules/ROOT/pages/drivers/analyze.adoc

```diff
@@ -0,0 +1,296 @@
+of having an abstract base class which is a union of all the variants.
+it must be downcast to the appropriate variant using the `is_<variant>` and `as_<variant>` methods.
+
+[source,python]
+----
+class PipelineStage(ABC):
+    def is_match(self) -> bool
+    def as_match(self) -> MatchStage
+    # is_insert, as_insert, is_select, as_select, ...
+
+class MatchStage(PipelineStage):
+    def block(self) -> ConjunctionID
+
+class SelectStage(PipelineStage):
+    def variables(self) -> Iterator[Variable]
+----
+
+Stages such as `Match` and `Insert` hold a `ConjunctionID`.
+This is an indirection which can be used
+to retrieve the actual conjunction using the `Pipeline.conjunction` method.
+
+[source,python]
+----
+#!test[]
+#{{
+include::{page-version}@core-concepts::example$driver/python_driver_usage.py[tag=import_and_constants]
+
+include::{page-version}@core-concepts::example$driver/python_driver_usage.py[tag=driver_create]
+
+include::{page-version}@core-concepts::example$driver/analyze_query.py[tag=analyze]
+
+#}}
+include::{page-version}@core-concepts::example$driver/analyze_query.py[tag=get_conjunction]
+----
+
+From the returned conjunction one can access the constraints,
+as well as the types inferred for the variables in those constraints.
+[source,python]
+----
+class Conjunction:
+    def constraints(self) -> Iterator[Constraint]
+    def annotated_variables(self) -> Iterator[Variable]
+    def variable_annotations(self, variable: Variable) -> Optional[VariableAnnotations]
+----
+
+=== Constraints
+Similar to stages, the `Constraint` instances returned by the `constraints()` method
+must be down-cast to the appropriate variant using the `is/as` methods.
+Sub-patterns such as `or`, `not`, and `try` are also constraints.
+These hold the `ConjunctionID`(s) of the nested conjunctions.
+
+[source,python]
+----
+class Constraint(ABC):
+    def is_isa(self) -> bool
+    def as_isa(self) -> Isa
+    # is_has, as_has, ...
+
+    def is_or(self) -> bool
+    def as_or(self) -> Or
+    # is_not, as_not, is_try, as_try
+
+class Isa(Constraint):
+    # <instance> isa(!) <type>
+    def instance(self) -> ConstraintVertex
+    def type(self) -> ConstraintVertex
+    def exactness(self) -> ConstraintExactness # isa or isa!
+
+# Has, ...
+
+class Or(Constraint):
+    def branches(self) -> Iterator[ConjunctionID]
+# Not, Try
+----
+
+To get the `isa` constraint from the first branch:
+[source,python]
+----
+#!test[]
+#{{
+include::{page-version}@core-concepts::example$driver/python_driver_usage.py[tag=import_and_constants]
+
+include::{page-version}@core-concepts::example$driver/python_driver_usage.py[tag=driver_create]
+
+include::{page-version}@core-concepts::example$driver/analyze_query.py[tags=analyze;get_conjunction]
+
+#}}
+include::{page-version}@core-concepts::example$driver/analyze_query.py[tag=get_first_branch_isa]
+----
+
+==== Constraint vertices
+Although constraints typically apply on variables,
```

Comment:
Quick explainer of why "vertex" makes sense - we can think of the query as a graph

Reply:
I don't think vertex is the ideal term but it's what we have in core. These are just variables which have a fixed value but I don't know of a term that generalises variables and constants. Constraint arguments? Sadly it's released so we're stuck with the term from core now.

Constraints are ideally independent of the graph view. If I could sneak vertex in without bringing the graph-view into this article, I'd prefer that.